pnet_datalink/src/linux.rs

```diff
@@ -97,19 +97,25 @@ impl Default for Config {
 
 /// Create a data link channel using the Linux's `AF_PACKET` socket type.
 #[inline]
-pub fn channel(network_interface: &NetworkInterface, config: Config) -> io::Result<super::Channel> {
-    let eth_p_all = 0x0003;
-    let (typ, proto) = match config.channel_type {
-        super::ChannelType::Layer2 => (libc::SOCK_RAW, eth_p_all),
-        super::ChannelType::Layer3(proto) => (libc::SOCK_DGRAM, proto),
+pub fn channel(
+    network_interface: &NetworkInterface,
+    config: Config,
+    #[cfg(feature = "psocket")]
+    socket: i32
+) -> io::Result<super::Channel> {
+    let (_typ, proto) = match config.channel_type {
+        super::ChannelType::Layer2 => (libc::SOCK_RAW, libc::ETH_P_ALL),
+        super::ChannelType::Layer3(proto) => (libc::SOCK_DGRAM, proto as i32),
     };
-    let socket = unsafe { libc::socket(libc::AF_PACKET, typ, proto.to_be() as i32) };
-    if socket == -1 {
-        return Err(io::Error::last_os_error());
-    }
-    let mut addr: libc::sockaddr_storage = unsafe { mem::zeroed() };
-    let len = network_addr_to_sockaddr(network_interface, &mut addr, proto as i32);
 
+    #[cfg(not(feature = "psocket"))]
```

Comment:
I don't think this should be feature gated - it should be configurable at runtime using the `Config` struct - if the socket is already set then we skip this setup.